From 3e325e3b94d0cbfb237e44b3bf1a3867fc279149 Mon Sep 17 00:00:00 2001
From: Adam Kaminski <adam@adamkaminski.com>
Date: Tue, 29 Sep 2026 15:30:08 -0400
Subject: [PATCH] fix(aws-vpc-move-ip): fix assumed-role credentials and
 diagnostics

Query STS credential fields explicitly and validate the returned values.
Preserve command output and exit status, and return OCF_ERR_GENERIC when
assuming the role fails.
Report resource-agent failures through ocf_exit_reason and clarify
diagnostic messages while preserving the original script formatting.
---
 heartbeat/aws-vpc-move-ip | 85 ++++++++++++++++++++++++---------------
 1 file changed, 53 insertions(+), 32 deletions(-)

diff --git a/heartbeat/aws-vpc-move-ip b/heartbeat/aws-vpc-move-ip
index 1348c59085..f5af390e1e 100755
--- a/heartbeat/aws-vpc-move-ip
+++ b/heartbeat/aws-vpc-move-ip
@@ -1,6 +1,5 @@
 #!/bin/sh
 #
-#
 # OCF resource agent to move an IP address within a VPC in the AWS
 #
 # Copyright (c) 2017 Markus Guertler (SUSE)
@@ -25,8 +24,6 @@
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write the Free Software Foundation,
 # Inc., 59 Temple Place - Suite 330, Boston MA 02111-1307, USA.
-#
-
 
 #######################################################################
 # Initialization:
@@ -65,7 +62,6 @@ OCF_RESKEY_awscli_timeout_default=""
 : ${OCF_RESKEY_awscli_timeout=${OCF_RESKEY_awscli_timeout_default}}
 #######################################################################
 
-
 USAGE="usage: $0 {start|stop|status|meta-data}";
 ###############################################################################
 
@@ -234,22 +230,43 @@ awscli cli-connect-timeout value
 END
 }
 
-
 execute_cmd_as_role(){
 	cmd=$1
 	role=$2
-	output="$($AWSCLI_CMD sts assume-role --role-arn $role --role-session-name AWSCLI-RouteTableUpdate --output=text)"
-	export AWS_ACCESS_KEY_ID="$(echo $output | awk -F" " '$4=="CREDENTIALS" {print $5}')"
-	export AWS_SECRET_ACCESS_KEY="$(echo $output | awk -F" " '$4=="CREDENTIALS" {print $7}')"
-	export AWS_SESSION_TOKEN="$(echo $output | awk -F" " '$4=="CREDENTIALS" {print $8}')"
 
-	#Execute command
-	ocf_log debug "Assumed Role ${role}"
+	output="$($AWSCLI_CMD sts assume-role \
+		--role-arn "$role" \
+		--role-session-name AWSCLI-RouteTableUpdate \
+		--query 'Credentials.[AccessKeyId,SecretAccessKey,SessionToken]' \
+		--output text)"
+	rc=$?
+	if [ "$rc" -ne 0 ]; then
+		unset output
+		ocf_exit_reason "Failed to assume role: ${role}"
+		return "$OCF_ERR_GENERIC"
+	fi
+
+	set -- $output
+	if [ "$#" -ne 3 ] || [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
+		unset output
+		ocf_exit_reason "Invalid credentials returned while assuming role: ${role}"
+		return "$OCF_ERR_GENERIC"
+	fi
+
+	export AWS_ACCESS_KEY_ID="$1"
+	export AWS_SECRET_ACCESS_KEY="$2"
+	export AWS_SESSION_TOKEN="$3"
+	unset output
+
+	# Execute command
+	ocf_log debug "assumed role: ${role}"
 	ocf_log debug "$($OCF_RESKEY_awscli sts get-caller-identity)"
 	ocf_log debug "executing command: $cmd"
 	response="$($cmd)"
-	unset output AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN
-	echo $response
+	rc=$?
+	unset AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN
+	printf '%s\n' "$response"
+	return "$rc"
 }
 
 ec2ip_set_address_param_compat(){
@@ -265,14 +282,14 @@ ec2ip_validate() {
 	done
 
 	if [ "x${OCF_RESKEY_auth_type}" = "xkey" ] && [ -z "$OCF_RESKEY_profile" ]; then
-		ocf_exit_reason "profile parameter not set"
+		ocf_exit_reason "Profile parameter is not set"
 		return $OCF_ERR_CONFIGURED
 	fi
 
 	if [ -n "$OCF_RESKEY_iflabel" ]; then
 		label=${OCF_RESKEY_interface}:${OFC_RESKEY_iflabel}
 		if [ ${#label} -gt 15 ]; then
-			ocf_exit_reason "Interface label [$label] exceeds maximum character limit of 15"
+			ocf_exit_reason "Interface label exceeds maximum character limit (15): $label"
 			exit $OCF_ERR_CONFIGURED
 		fi
 	fi
@@ -283,7 +300,7 @@ ec2ip_validate() {
 	[ $? -ne 0 ] && exit $OCF_ERR_GENERIC
 
 	if [ -z "${EC2_INSTANCE_ID}" ]; then
-		ocf_exit_reason "Instance ID not found. Is this a EC2 instance?"
+		ocf_exit_reason "Failed to determine instance ID. Is this an EC2 instance?"
 		return $OCF_ERR_GENERIC
 	fi
 
@@ -302,7 +319,7 @@ ec2ip_monitor() {
 
 	if ocf_is_true ${OCF_RESKEY_monapi} || [ "$__OCF_ACTION" = "start" ] || ocf_is_probe; then
 		for rtb in $(echo $OCF_RESKEY_routing_table | sed -e 's/,/ /g'); do
-			ocf_log info "monitor: check routing table (API call) - $rtb"
+			ocf_log info "monitor: checking route table (API call): $rtb"
 			if [ -z "${OCF_RESKEY_routing_table_role}" ]; then
 				cmd="$AWSCLI_CMD --output text ec2 describe-route-tables --route-table-ids $rtb --query RouteTables[*].Routes[?DestinationCidrBlock=='$OCF_RESKEY_ip/32'].$OCF_RESKEY_lookup_type"
 				ocf_log debug "executing command: $cmd"
@@ -311,13 +328,14 @@ ec2ip_monitor() {
 				cmd="$OCF_RESKEY_awscli $region_opt --output text ec2 describe-route-tables --route-table-ids $rtb --query RouteTables[*].Routes[?DestinationCidrBlock=='$OCF_RESKEY_ip/32'].$OCF_RESKEY_lookup_type"
 				ROUTE_TO_INSTANCE="$(execute_cmd_as_role "$cmd" $OCF_RESKEY_routing_table_role)"
 			fi
-			ocf_log debug "Overlay IP is currently routed to ${ROUTE_TO_INSTANCE}"
+
 			if [ -z "$ROUTE_TO_INSTANCE" ]; then
 				ROUTE_TO_INSTANCE="<unknown>"
 			fi
+			ocf_log debug "overlay IP address is currently routed to ${ROUTE_TO_INSTANCE}"
 
 			if [ "$EC2_ID" != "$ROUTE_TO_INSTANCE" ]; then
-				ocf_log warn "not routed to this $RESOURCE_TYPE ($EC2_ID) but to $RESOURCE_TYPE $ROUTE_TO_INSTANCE on $rtb"
+				ocf_log warn "Overlay IP is not routed to $RESOURCE_TYPE ($EC2_ID) but to $RESOURCE_TYPE $ROUTE_TO_INSTANCE on $rtb"
 				MON_RES="$MON_RES $rtb"
 			fi
 			sleep 1
@@ -336,16 +354,15 @@ ec2ip_monitor() {
 	RESULT=$($cmd | grep "$OCF_RESKEY_ip")
 	if [ -z "$RESULT" ]; then
 		if [ "$__OCF_ACTION" = "monitor" ] && ! ocf_is_probe; then
-			level="error"
+			ocf_exit_reason "IP address $OCF_RESKEY_ip not assigned to interface $OCF_RESKEY_interface"
 		else
-			level="info"
+			ocf_log info "IP address $OCF_RESKEY_ip not assigned to interface $OCF_RESKEY_interface"
 		fi
 
-		ocf_log "$level" "IP $OCF_RESKEY_ip not assigned to interface $OCF_RESKEY_interface"
 		return $OCF_NOT_RUNNING
 	fi
 
-	ocf_log debug "route in VPC and address assigned"
+	ocf_log debug "route in VPC and IP address assigned"
 	return $OCF_SUCCESS
 }
 
@@ -433,48 +450,50 @@ ec2ip_stop() {
 
 	ec2ip_monitor
 	if [ $? = $OCF_NOT_RUNNING ]; then
-		ocf_log info "EC2: Address $OCF_RESKEY_ip already down"
+		ocf_log info "EC2: IP address $OCF_RESKEY_ip is already down"
 		return $OCF_SUCCESS
 	fi
 
 	ec2ip_drop
 	if [ $? != $OCF_SUCCESS ]; then
+		ocf_log warn "EC2: Unable to remove local IP address $OCF_RESKEY_ip from interface $OCF_RESKEY_interface"
 		return $OCF_ERR_GENERIC
 	fi
 
 	ec2ip_monitor
 	if [ $? != $OCF_NOT_RUNNING ]; then
-		ocf_log error "EC2: Couldn't bring down IP address $OCF_RESKEY_ip on interface $OCF_RESKEY_interface."
+		ocf_exit_reason "Unable to remove local IP address $OCF_RESKEY_ip from interface $OCF_RESKEY_interface"
 		return $OCF_ERR_GENERIC
 	fi
 
-	ocf_log info "EC2: Successfully brought down $OCF_RESKEY_ip"
+	ocf_log info "EC2: Successfully brought down IP address $OCF_RESKEY_ip"
 	return $OCF_SUCCESS
 }
 
 ec2ip_start() {
-	ocf_log info "EC2: Moving IP address $OCF_RESKEY_ip to this host by adjusting routing table $OCF_RESKEY_routing_table"
+	ocf_log info "EC2: Moving IP address $OCF_RESKEY_ip to this instance and adjusting VPC route table $OCF_RESKEY_routing_table"
 
 	ec2ip_monitor
 	if [ $? = $OCF_SUCCESS ]; then
-		ocf_log info "EC2: $OCF_RESKEY_ip already started"
+		ocf_log info "EC2: IP address $OCF_RESKEY_ip is already started"
 		return $OCF_SUCCESS
 	fi
 
-	ocf_log info "EC2: Adjusting routing table and locally configuring IP address"
+	ocf_log info "EC2: Adjusting route and configuring IP address"
 	ec2ip_get_and_configure
 	rc=$?
 	if [ $rc != $OCF_SUCCESS ]; then
-		ocf_log error "Received $rc from 'aws'"
+		ocf_exit_reason "EC2: Unable to configure route or IP address (rc: $rc)"
 		return $OCF_ERR_GENERIC
 	fi
 
 	ec2ip_monitor
 	if [ $? != $OCF_SUCCESS ]; then
-		ocf_log error "EC2: IP address couldn't be configured on this host (IP: $OCF_RESKEY_ip, Interface: $OCF_RESKEY_interface)"
+		ocf_exit_reason "EC2: Unable to verify VPC routing and IP address $OCF_RESKEY_ip on interface $OCF_RESKEY_interface"
 		return $OCF_ERR_GENERIC
 	fi
 
+	ocf_log info "EC2: Successfully configured IP address $OCF_RESKEY_ip on this instance and adjusted VPC route table $OCF_RESKEY_routing_table"
 	return $OCF_SUCCESS
 }
 
@@ -496,7 +515,7 @@ case $__OCF_ACTION in
 esac
 
 if ! ocf_is_root; then
-	ocf_log err "You must be root for $__OCF_ACTION operation."
+	ocf_exit_reason "Must be root for $__OCF_ACTION operation"
 	exit $OCF_ERR_PERM
 fi
 
@@ -538,3 +557,5 @@ case $__OCF_ACTION in
 		exit $OCF_ERR_UNIMPLEMENTED
 		;;
 esac
+
+# vim: set ts=4 sw=4 sts=4 noet tw=80 :
